src/domain_bridge/domain_bridge.cpp

```diff
@@ -0,0 +1,210 @@
+  }
+
+  void create_subscription(
+    rclcpp::Node::SharedPtr node,
+    SubscriptionMap & map,
+    std::shared_ptr<GenericPublisher> publisher,
+    const std::string & topic_name,
+    const rosidl_message_type_support_t & typesupport_handle)
+  {
+    size_t domain_id = node->get_node_options().context()->get_domain_id();
+    auto domain_id_topic_pair = std::make_pair(domain_id, topic_name);
+    if (map.find(domain_id_topic_pair) != map.end()) {
+      throw BridgeExistsError(
+              node->get_node_options().context()->get_domain_id(),
+              topic_name
+      );
+    }
+
+    // Create subscription
+    auto subscription = std::make_shared<GenericSubscription>(
+      node->get_node_base_interface().get(),
+      typesupport_handle,
+      topic_name,
+      rclcpp::QoS(10),
+      [publisher](std::shared_ptr<rclcpp::SerializedMessage> msg) {
+        // Publish message into the other domain
+        auto serialized_data_ptr = std::make_shared<rcl_serialized_message_t>(
+          msg->get_rcl_serialized_message());
+        publisher->publish(serialized_data_ptr);
+      });
+    node->get_node_topics_interface()->add_subscription(subscription, nullptr);
+    map[domain_id_topic_pair] = subscription;
+  }
+
+  void bridge_topic(
+    const std::string & topic,
+    const std::string & type,
+    size_t from_domain_id,
+    size_t to_domain_id,
+    const TopicBridgeOptions & options)
+  {
+    // TODO(jacobperron): Do something with options.callback_group()
+    (void)options;
+
+    rclcpp::Node::SharedPtr from_domain_node = get_node_for_domain(from_domain_id);
+    rclcpp::Node::SharedPtr to_domain_node = get_node_for_domain(to_domain_id);
+
+    // Get typesupport
+    auto typesupport_library = rosbag2_cpp::get_typesupport_library(
+      type, "rosidl_typesupport_cpp");
+    auto typesupport_handle = rosbag2_cpp::get_typesupport_handle(
+      type, "rosidl_typesupport_cpp", typesupport_library);
+
+    // Create publisher for the 'to_domain' and subscription for the 'from_domain'
+    // The publisher should be created first so it is available to the subscription callback
+    auto publisher = this->create_publisher(
+      to_domain_node, this->publisher_map_, topic, *typesupport_handle);
+    this->create_subscription(
+      from_domain_node, this->subscription_map_, publisher, topic, *typesupport_handle);
+  }
+
+  void add_to_executor(std::shared_ptr<rclcpp::Executor> executor)
+  {
+    for (const auto & domain_id_node_pair : this->node_map_) {
+      executor->add_node(domain_id_node_pair.second);
+    }
+  }
+
+  /// Map of domain IDs to ROS nodes
+  NodeMap node_map_;
+
+  /// Map of (domain id, topic name) to publishers
+  PublisherMap publisher_map_;
+
+  /// Map of (domain_id, topic name) to subscriptions
+  SubscriptionMap subscription_map_;
+};  // class DomainBridgeImpl
+
+DomainBridge::DomainBridge()
+: impl_(std::make_unique<DomainBridgeImpl>())
+{}
+
+DomainBridge::~DomainBridge()
+{}
```

Comment:
```suggestion
DomainBridge::~DomainBridge() = default;
```

iiuc you cannot move this to the header file because `DomainBridgeImpl` was forward declared there, but `= default` is still valid in the source file.

Reply:
a559080